Add new module for distributing tokens from COW DAO

fedgiac opened the pull request:
It's being discussed to have the DAO stream COW token to third party as part of a DAO proposal.

Using the code in this repository would be a simple solution: by design, it allows to dynamically allocate some COW to users linearly for a specific amount of time.
It was originally designed to claim vCOW into COW before selling, but it can be used as well in the case where there's no vCOW tokens but only COW, as it's the case for the DAO.

### Changes

- Add new deployment script with different parameters.
- Update `hardhat-deploy` to the latest version.
- Extend the existing unit test to work with both deployments. The only expected logic change is that the distributor balance is compared in aggregate (vCOW+COW) rather than each balance independently. This is because the DAO only holds COW and no vCOW, so the relative balance change is different in the two cases. 
- Add Sepolia for testing.

My thanks to wighawag for his [help with hardhat-deploy](https://github.com/wighawag/hardhat-deploy/issues/584)!

### Test Plan

New mainnet forked test (based on the already existing test).

Also, try to deploy the modules on Sepolia following the instructions. You can try to execute it partially: first only the first deployment and then the second one, to make sure that the existing contract is detected and reused.

<details><summary>Constants needed for Sepolia</summary> 

```
export const COW_TOKEN = "0x0625aFB445C3B6B7B929342a04A22599fd5dBB59";
export const VIRTUAL_COW_TOKEN = "0x21d06A222BBB94Ec1406a0a8Ba86B4D761bc9864";
```
</details>

fedgiac committed:
Update hardhat-deploy

fedgiac committed:
Add module for CoW DAO

fedgiac committed:
Update description

fedgiac committed:
Separate deployment names from contract name (and delete unused contract)

kaze-cow commented on README.md:
```diff
@@ -1,19 +1,23 @@
-When a claim is created, an address is assigned a specific vCOW amount that will be vested linearly in the chosen amount of time.
+The allocation module distributes COW and vCOW tokens from a Safe multisig.
+
+This module has two known use cases:
```
when we say "known" use cases, does that mean they are both currently in use as usecases, or just things we might do in the future?

fedgiac replied:
Clarified: ed830f76ba9d460a03c1e766c46a41699a7272bf

kaze-cow commented on test/mainnet/team-controller-module.ts:
```diff
@@ -71,175 +74,198 @@ async function execSafeTransaction(
+    });
   });
+}
+
+// The involved safes need to have enough COW/vCOW to make the test pass at this block.
+const MAINNET_BLOCK = 23890068;
+
+testModule({
+  name: "team allocation",
+  deployment: TEAM_ALLOCATION_DEPLOYMENT_NAME,
+  safe: TEAM_CONTROLLER_SAFE,
+  forkBlock: MAINNET_BLOCK,
+});
+
+testModule({
+  name: "COW DAO allocation",
+  deployment: DAO_ALLOCATION_DEPLOYMENT_NAME,
+  safe: COW_DAO,
+  forkBlock: MAINNET_BLOCK,
 });
```
worth mentioning this block right here is the only effective change I can see for this file. Most of the other changes are just renames and adding another condition for getting the total COW + vCOW

anxolin replied:
Yep, the formatting made it a bit harder to know the changes

kaze-cow commented on README.md:
````diff
@@ -65,11 +69,11 @@ REPORT_GAS=1 yarn test
 ```
 
-It will be associated with the mainnet team controller address automatically.
+It will be associated with the mainnet safe address automatically.
````
maybe worth mentioning how this safe address is set up. I can see its configured in `constants.ts`

fedgiac replied:
Linked: 59b17d870c48a7f925613cbd25940dd230164b43

kaze-cow reviewed (approved):
besides the core specifically mentioned changes, a lot of sort of unrelated changes (ex. renaming variables) which makes it a bit more difficult to understand what is actually happening.

otherwise we good

kaze-cow commented on src/ts/constants.ts:
```diff
@@ -2,3 +2,4 @@ export const TEAM_CONTROLLER_SAFE =
   "0xca07EaA4253638D286caD71CBcEec11803F2709A";
 export const COW_TOKEN = "0xDEf1CA1fb7FBcDC777520aa7f396b4E015F497aB";
 export const VIRTUAL_COW_TOKEN = "0xD057B63f5E69CF1B929b356b579Cba08D7688048";
+export const COW_DAO = "0xcA771eda0c70aA7d053aB1B25004559B918FE662";
```
for consistency would be good to add `_SAFE` postfix to this var name to clarify it is also a safe like `TEAM_CONTROLLER_SAFE`

fedgiac replied:
Makes sense: 6339a05dc4c29135ba4104ff79c23b4ac249165e

anxolin commented on hardhat.config.ts:
```diff
@@ -98,6 +98,11 @@ export default {
       chainId: 1,
     },
+    sepolia: {
+      url: `https://sepolia.infura.io/v3/${INFURA_KEY}`,
```
I see you replaced `INFURA_KEY` with `NODE_URL` above. Would the change affect this file?

fedgiac replied:
The way the struct is defined makes it so that `NODE_URL` replaces this `url` if the env exists.
The reason I changed the readme is that I don't think we want to endorse using Infura anymore since we've all gotten used to specifying a node URL. Infura should still work as before anyway.

anxolin commented on src/deploy/002_CowDaoAllocationModule.ts:
```diff
@@ -0,0 +1,44 @@
+  CONTRACT_NAME,
+} from "../ts";
+
+const deployCowDaoAuthenticator: DeployFunction = async function ({
```
This function is the same as the one as 001_AllocationModule, just the contract name changes. 

I would suggest to refactor into a function. This way, code is dryer but also is more clear that we deploy the contract twice, one for the initial team allocation, another for cow dao allocations and only the controller and name changes

fedgiac replied:
Done! (Three commits because there were a few issues, 7f23e49a2ec16e477dbab442bc136fd030dc8b7a d4bcc130304592881ebb8b6547d92a6a298d73b3 2d6ba261e82388e378044b6e664d8bf9ce49521f.)

I tested it worked by deploying this on Sepolia (with minimal changes to the contract to avoid reverting on initialization):
<img width="1514" height="357" alt="image" src="https://github.com/user-attachments/assets/2aac3b6a-cb15-438f-91ed-51639ddda8b9" />

Networks.json includes both deployments:
<img width="834" height="358" alt="image" src="https://github.com/user-attachments/assets/fc2696f7-a2be-42fb-8f28-6b5cddd1bb85" />

fedgiac committed:
Clarify usage

fedgiac committed:
Mention constants in readme

fedgiac committed:
COW_DAO → COW_DAO_SAFE

fedgiac committed:
Move deployed code to dedicated function

fedgiac committed:
Move deployment function to deployment files to work around a typing issue

CI was reverting while the function was located in the library files.
Moving it to the deployment files is a hack that fixes the issue.
I'm not really sure what's going on, but it looks like the type I need to use comes from type extensions and `declare module` so I'm not sure how to access them directly.

Relevant issue link:
https://github.com/wighawag/hardhat-deploy-ethers/issues/18

fedgiac committed:
Fix missing variable deployment name
